Source/WebCore/css/parser/CSSParserImpl.cpp

```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+            values.append(tagInteger);
+            if (values.size() > maxValues)
+                return { };
+        }
+        if (values.isEmpty())
+            return { };
+        
+        return { Tag { name.toString(), values } };
+    };
+
+    Vector<Tag> tags;
+    while (!range.atEnd()) {
+        switch (range.peek().type()) {
+        case WhitespaceToken:
+        case SemicolonToken:
+            range.consume();
+            break;
+        case IdentToken: {
+            const CSSParserToken* declarationStart = &range.peek();
+
+            while (!range.atEnd() && range.peek().type() != SemicolonToken)
+                range.consumeComponentValue();
+
+            if (auto tag = consumeTag(range.makeSubRange(declarationStart, &range.peek()), maxValues))
+                tags.append(*tag);
+
+            break;
+        }
+        default: // Parse error, unexpected token in declaration list
+            while (!range.atEnd() && range.peek().type() != SemicolonToken)
+                range.consumeComponentValue();
+            break;
+        }
+    }
+    return StyleRuleFontFeatureValuesBlock::create(type, tags);
+}
+RefPtr<StyleRuleFontFeatureValues> CSSParserImpl::consumeFontFeatureValuesRule(CSSParserTokenRange prelude, CSSParserTokenRange block)
+{
+    // @font-feature-values <family-name># { <declaration-list> }
+
+    auto fontFamilies = CSSPropertyParserHelpers::consumeFamilyNameList(prelude);
+    if (fontFamilies.isEmpty() || !prelude.atEnd())
+        return nullptr;
+    
+    Vector<RefPtr<StyleRuleBase>> rules;
+    consumeRuleList(block, FontFeatureValuesRuleList, [&rules](RefPtr<StyleRuleBase> rule) {
+        rules.append(rule);
+    });
+    rules.shrinkToFit();
+    
+    // Convert block rules to value (remove duplicate...etc)
+    FontFeatureValuesValue fontFeaturesValuesvalue;
+    auto currentTagsForType = [&fontFeaturesValuesvalue](FontFeatureValuesType type) -> Tags& {
+        switch (type) {
+        case FontFeatureValuesType::Styleset:
+            return fontFeaturesValuesvalue.styleset;
+        case FontFeatureValuesType::Stylistic:
+            return fontFeaturesValuesvalue.stylistic;
+        case FontFeatureValuesType::CharacterVariant:
+            return fontFeaturesValuesvalue.characterVariant;
+        case FontFeatureValuesType::Swash:
+            return fontFeaturesValuesvalue.swash;
+        case FontFeatureValuesType::Ornaments:
+            return fontFeaturesValuesvalue.ornaments;
+        case FontFeatureValuesType::Annotation:
+            return fontFeaturesValuesvalue.annotation;
+        }
+    };
+
+    auto upsertTag = [&currentTagsForType](FontFeatureValuesType type, Tag tag) {
+        auto& tags = currentTagsForType(type);
+        tags.set(tag.first, tag.second);
+    };
+    
+    for (auto block : rules) {
+        if (!block)
+            continue;
+        if (!block->isFontFeatureValuesBlockRule())
+            continue;
```

Comment:
Should this be an ASSERT()? (if yes, the downcast will do the assert for you and you can just delete these 2 lines)